gufe/storage/pseudodirectory.py

```diff
@@ -0,0 +1,156 @@
+        write to prevent NEW files from being written within this shared
+        directory. NOTE: This will not prevent overwrite of existing files,
+        in scratch space, but it will prevent changed files from uploading
+        to the external storage.
+    """
+    def __init__(
+        self,
+        scratch: PathLike,
+        external: ExternalStorage,
+        prefix: str,
+        *,
+        holding: PathLike = Path(".holding"),
+        delete_holding: bool = True,
+        read_only: bool = False,
+    ):
+        self.external = external
+        self.scratch = Path(scratch)
+        self.prefix = Path(prefix)
+        self.read_only = read_only
+        self.delete_holding = delete_holding
+        self.holding = holding
+
+        self.registry : set[SharedPath] = set()
+        # NOTE: the fact that we use $SCRATCH/$HOLDING/$PREFIX instead of
+        # $SCRATCH/$PREFIX/$HOLDING is important for 2 reasons:
+        # 1. This doesn't take any of the user's namespace from their
+        #    $SCRATCH/$PREFIX directory.
+        # 2. This allows us to easily use an external FileStorage where the
+        #    external storage is exactly the same as this local storage,
+        #    meaning that copies to/from the external storage are no-ops.
+        #    Use FileStorage(scratch / holding) for that.
+        self.shared_dir = self.scratch / holding / prefix
+        self.shared_dir.mkdir(exist_ok=True, parents=True)
+
+    def get_other_shared_dir(self, prefix, delete_holding=None):
+        """Get a related unit's shared directory.
+        """
+        if delete_holding is None:
+            delete_holding = self.delete_holding
+
+        return SharedRoot(
+            scratch=self.scratch,
+            external=self.external,
+            prefix=prefix,
+            holding=self.holding,
+            delete_holding=delete_holding,
+            read_only=True,
+        )
+
+    def transfer_holding_to_external(self):
+        """Transfer all objects in the registry to external storage"""
+        if self.read_only:
+            logging.debug("Read-only: Not transfering to external storage")
+            return  # early exit
+
+        for obj in self.registry:
+            path = Path(obj)
+            if not path.exists():
+                logging.info(f"Found nonexistent path {path}, not "
+                             "transfering to external storage")
+            elif path.is_dir():
+                logging.debug(f"Found directory {path}, not "
+                             "transfering to external storage")
+            else:
+                logging.info(f"Transfering {path} to external storage")
+                self.external.store_path(obj.label, path)
+
+    def __del__(self):
+        # take everything in self.shared_dir and write to it shared; keeping
+        # our prefix
+        self.transfer_holding_to_external()
+        if self.delete_holding:
+            shutil.rmtree(self.shared_dir)
```

Comment:
iirc `__del__` isn't called on `del obj` but instead when garbage collection happens.  Maybe instead make `transfer_holding_to_external` always explicit and don't rely on garbage collection related things

Reply:
works for me ... it's an executor-level thing anyway, and I don't mind asking the executor writer to think about how files get moved around (don't want to ask the protocol writer to think about that)

Reply:
This started out as a context manager. I switched it away because it makes actual usage kind of weird -- the entirety of the execution block where this is used gets indented by one level, which I don't particularly like. It gets worse if you have one of these for `shared` and one for `permanent` -- there's a syntax for setting multiple context managers at once, but I've also always found that kind of weird.

But I can switch it back to a context manager if you want 🤷‍♂️